protocols/gossipsub/src/behaviour.rs

```diff
@@ -0,0 +1,1165 @@
+        self.join(&topic_hash);
+        info!("Subscribed to topic: {}", topic);
+        true
+    }
+
+    /// Unsubscribes from a topic.
+    ///
+    /// Returns true if we were subscribed to this topic.
+    pub fn unsubscribe(&mut self, topic: Topic) -> bool {
+        debug!("Unsubscribing from topic: {}", topic);
+        let topic_hash = &self.topic_hash(topic);
+
+        if self.mesh.get(topic_hash).is_none() {
+            debug!("Already unsubscribed to topic: {:?}", topic_hash);
+            // we are not subscribed
+            return false;
+        }
+
+        // announce to all peers in the topic
+        if let Some(peer_list) = self.topic_peers.get(topic_hash) {
+            let event = Arc::new(GossipsubRpc {
+                messages: Vec::new(),
+                subscriptions: vec![GossipsubSubscription {
+                    topic_hash: topic_hash.clone(),
+                    action: GossipsubSubscriptionAction::Unsubscribe,
+                }],
+                control_msgs: Vec::new(),
+            });
+            for peer in peer_list {
+                debug!("Sending UNSUBSCRIBE to peer: {:?}", peer);
+                self.events.push_back(NetworkBehaviourAction::SendEvent {
+                    peer_id: peer.clone(),
+                    event: event.clone(),
+                });
+            }
+        }
+
+        // call LEAVE(topic)
+        // this will remove the topic from the mesh
+        self.leave(&topic_hash);
+
+        info!("Unsubscribed from topic: {:?}", topic_hash);
+        true
+    }
+
+    /// Publishes a message to the network.
+    pub fn publish(&mut self, topic: &Topic, data: impl Into<Vec<u8>>) {
+        self.publish_many(iter::once(topic.clone()), data)
+    }
+
+    /// Publishes a message with multiple topics to the network.
+    pub fn publish_many(
+        &mut self,
+        topic: impl IntoIterator<Item = Topic>,
+        data: impl Into<Vec<u8>>,
+    ) {
+        let message = GossipsubMessage {
+            source: self.local_peer_id.clone(),
+            data: data.into(),
+            // If the sequence numbers are predictable, then an attacker could flood the network
+            // with packets with the predetermined sequence numbers and absorb our legitimate
+            // messages. We therefore use a random number.
+            // TODO: Check if the random sequence numbers causes issues with other clients.
+            // To be interoperable with the go-implementation this is treated as a 64-bit
+            // big-endian uint.
+            sequence_number: rand::random::<[u8; 8]>().to_vec(),
+            topics: topic.into_iter().map(|t| self.topic_hash(t)).collect(),
+        };
+
+        debug!("Publishing message: {:?}", message.id());
+
+        // forward the message to mesh and floodsub peers
+        let local_peer_id = self.local_peer_id.clone();
+        self.forward_msg(message.clone(), &local_peer_id);
+
+        let mut recipient_peers = HashMap::new();
```

Comment:
```suggestion
        let mut recipient_peers = HashSet::new();
```

A HashSet would be more idiomatic here, right?

Reply:
Yes of course. Nice catch!